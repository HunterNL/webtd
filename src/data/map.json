{
    "entities" : [
        {
            "type": "end",
            "id": 0
        },
        {
            "type" : "end",
            "id": 1
        },
        {
            "type" : "track",
            "id": 2,
            "length" : 500,
            "ends" : [0,1]
        }
    ]
}