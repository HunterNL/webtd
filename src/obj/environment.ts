import { Track, isTrack } from "./track";
import { Train, isTrain } from "./train";
import { isObject } from "../util/isObject";
import { Identifiable, isIdentifiable } from "../interfaces/id";
import { Entity } from "../interfaces/entity";


export type EnvironmentSave = {
    entities: Entity[]
}

export function isValidSafeData(any: any): any is EnvironmentSave {
    return Array.isArray(any.entities) && any.entities.every(isIdentifiable);
}

export type Environment=  {
    tracks: Track[],
    trains: Train[]
}

export function loadEnvironment(map: unknown): Environment {
    if(!isValidSafeData(map)) {
        throw new Error("Invalid save");
    }

    const {entities} = map;

    const tracks: Track[] = entities.filter(isTrack);
    const trains: Train[] = entities.filter(isTrain);

    

    return {
        tracks,
        trains,
    }
}