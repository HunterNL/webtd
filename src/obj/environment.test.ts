import map from "../data/map.json";
import { loadEnvironment } from "./environment";

test("Env loader",function() {
    expect(loadEnvironment(map)).toMatchSnapshot();
})