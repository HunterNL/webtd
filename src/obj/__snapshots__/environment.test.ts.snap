// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Env loader 1`] = `
Object {
  "tracks": Array [],
  "trains": Array [],
}
`;
